plugins/out_stackdriver/stackdriver_resource_types.c

```diff
@@ -0,0 +1,139 @@
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+#include <fluent-bit/flb_utils.h>
+#include <fluent-bit/flb_sds.h>
+#include <fluent-bit/flb_kv.h>
+
+#include "stackdriver.h"
+#include "stackdriver_resource_types.h"
+
+static const struct resource_type resource_types[] = {
+    { 
+        .id = RESOURCE_TYPE_K8S,
+        .resources = {"k8s_container", "k8s_node", "k8s_pod"},
+        .required_labels = {"cluster_name", "location"}
+    },
+    { 
+        .id = RESOURCE_TYPE_GENERIC_NODE,
+        .resources = {"generic_node"},
+        .required_labels = {"location", "namespace", "node_id"}
+    },
+    { 
+        .id = RESOURCE_TYPE_GENERIC_TASK,
+        .resources = {"generic_task"},
+        .required_labels = {"location", "namespace", "job", "task_id"}
+    }
+};
+
+static char **get_required_labels(int resource_type)
+{
+    int i;
+    int len;
+
+    len = sizeof(resource_types) / sizeof(resource_types[0]);
+    for(i = 0; i < len; i++) {
+        if (resource_types[i].id == resource_type) {
+            return (char **) resource_types[i].required_labels;
+        }
+    }
+    return NULL;
+}
+
+/*
+ *   set_resource_type():
+ * - Iterates through resource_types that are set up for validation and sets the 
+ *   resource_type if it matches one of them.
+ */
+void set_resource_type(struct flb_stackdriver *ctx)
+{
+    int i;
+    int j;
+    int len;
+    char *resource;
+    struct resource_type resource_type;
+
+    len = sizeof(resource_types) / sizeof(resource_types[0]);
+    for(i = 0; i < len; i++) {
+        resource_type = resource_types[i];
+        for(j = 0; j < MAX_RESOURCE_ENTRIES; j++) {
+            if (resource_type.resources[j] != NULL) {
+                resource = resource_type.resources[j];
+                if (flb_sds_cmp(ctx->resource, resource, strlen(resource)) == 0) {
+                    ctx->resource_type = resource_type.id;
+                    return;
+                }
+            }
+        }
+    }
+}
+
+/*
+ *   resource_api_has_required_labels():
+ * - Determines if all required labels for the set resource type are present as
+ *   keys on the resource labels key-value pairs.
+ */
+int resource_api_has_required_labels(struct flb_stackdriver *ctx)
+{
+    struct mk_list *head;
+    struct flb_hash *ht;
+    struct flb_kv *label_kv;
+    char** required_labels;
+    int i;
+    int found;
+    void *tmp_buf;
+    size_t tmp_size;
+
+    if (mk_list_size(&ctx->resource_labels_kvs) == 0) {
+        return FLB_FALSE;
+    }
+
+    required_labels = get_required_labels(ctx->resource_type);
+    if (required_labels == NULL) {
+        return FLB_FALSE;
```

Comment:
The scenario that the `required_labels` will be NULL is that the resource_type is not supported. We don't catch or print out this error here or the function it returns back https://github.com/fluent/fluent-bit/pull/5623/files#diff-341af19a69d101080d3a956d73990e5fcfb9834de72e941ddbab62fe930dd565R460. We should either warn or error users about that the resource type is not supported.

Reply:
This current behavior is just for validation - note that we do not set the `should_skip_resource_labels_api` flag in this case. The intention is that this allows us to support new resource types using this API without having the friction of enforcing validation on the set labels from the get-go.

I think a warning probably is appropriate in this case though, so I have added one.

Reply:
SG. Thanks for adding the warning.